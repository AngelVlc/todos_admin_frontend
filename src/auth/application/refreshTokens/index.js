export * from "./GetRefreshTokensUseCase";
export * from "./DeleteSelectedRefreshTokensUseCase";
