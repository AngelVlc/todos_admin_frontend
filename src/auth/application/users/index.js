export * from "./GetUserByIdUseCase";
export * from "./DeleteUserByIdUseCase";
export * from "./GetUsersUseCase";
export * from "./CreateUserUseCase";
export * from "./UpdateUserUseCase";
