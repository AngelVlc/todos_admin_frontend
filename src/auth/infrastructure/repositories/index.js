export * from './RefreshTokensRepository';
export * from './UsersRepository';
