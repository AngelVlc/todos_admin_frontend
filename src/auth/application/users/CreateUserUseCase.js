import { BaseUseCase } from "../../../shared/domain/BaseUseCase";
import { UsersRepository } from "../../infrastructure/repositories";

export class CreateUserUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new UsersRepository();

    return new CreateUserUseCase({ repository });
  }

  async execute(user) {
    return await this._repository.create(user);
  }
}
