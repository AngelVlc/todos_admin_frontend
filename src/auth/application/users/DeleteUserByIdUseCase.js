import { BaseUseCase } from "../../../shared/domain/BaseUseCase";
import { UsersRepository } from "../../infrastructure/repositories";

export class DeleteUserByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new UsersRepository();

    return new DeleteUserByIdUseCase({ repository });
  }

  async execute(id) {
    return await this._repository.deleteById(id);
  }
}
