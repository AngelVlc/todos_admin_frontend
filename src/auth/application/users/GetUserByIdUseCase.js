import { BaseUseCase } from "../../../shared/domain/BaseUseCase";
import { User } from "../../domain";
import { UsersRepository } from "../../infrastructure/repositories";

export class GetUserByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new UsersRepository();

    return new GetUserByIdUseCase({ repository });
  }

  async execute(id) {
    const result = await this._repository.getById(id);

    return new User(result);
  }
}
