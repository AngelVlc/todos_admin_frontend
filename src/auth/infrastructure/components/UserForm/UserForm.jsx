import React, { useState, useContext } from "react";
import { useHistory } from "react-router-dom";
import { Formik, Form, Field, ErrorMessage } from "formik";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import * as Yup from "yup";
import { CreateUserUseCase } from "../../../application/users/CreateUserUseCase";
import { UpdateUserUseCase } from "../../../application/users/UpdateUserUseCase";

export const UserForm = (props) => {
  let history = useHistory();
  const { useCaseFactory } = useContext(AppContext);

  const [pageState] = useState(props.user);

  const onSubmit = async (user) => {
    let useCase;
    if (props.user?.id === undefined) {
      useCase = useCaseFactory.get(CreateUserUseCase);
    } else {
      useCase = useCaseFactory.get(UpdateUserUseCase);
    }
    const result = await useCase.execute(user);
    if (result) {
      history.push("/users");
    }
  };

  return (
    <Formik
      enableReinitialize={true}
      initialValues={pageState}
      validationSchema={Yup.object({
        name: Yup.string().required("Required"),
      })}
      onSubmit={onSubmit}
    >
      <Form>
        <div className="field">
          <label className="label" htmlFor="name">
            Name
          </label>
          <div className="control">
            <Field name="name" type="text" data-testid="name" autoFocus />
          </div>
          <p className="help is-danger" data-testid="nameErrors">
            <ErrorMessage name="name" />
          </p>
        </div>

        <div className="field">
          <label className="label" htmlFor="password">
            Password
          </label>
          <div className="control">
            <Field name="password" type="password" data-testid="password" />
          </div>
        </div>

        <div className="field">
          <label className="label" htmlFor="confirmPassword">
            Confirm Password
          </label>
          <div className="control">
            <Field
              name="confirmPassword"
              type="password"
              data-testid="confirmPassword"
            />
          </div>
        </div>

        <div className="field">
          <label className="label" htmlFor="isAdmin">
            Is Admin
          </label>
          <div className="control">
            <Field name="isAdmin" type="checkbox" data-testid="isAdmin" />
          </div>
        </div>

        <div className="field is-grouped">
          <div className="control">
            <button className="button" data-testid="submit" type="submit">
              {props.user?.id ? "UPDATE" : "CREATE"}
            </button>
          </div>
          <div className="control">
            <button
              className="button"
              data-testid="cancel"
              type="button"
              onClick={() => history.push("/users")}
            >
              CANCEL
            </button>
          </div>
          {props.children}
        </div>
      </Form>
    </Formik>
  );
};
