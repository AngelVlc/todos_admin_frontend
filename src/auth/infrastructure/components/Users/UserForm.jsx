import React, { useEffect, useState, useContext } from "react";
import { useHistory } from "react-router-dom";
import { Formik, Form, Field, ErrorMessage } from "formik";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import * as Yup from "yup";
import { User } from "../../../domain";
import { CreateUserUseCase } from "../../../application/users/CreateUserUseCase";
import { UpdateUserUseCase } from "../../../application/users/UpdateUserUseCase";

export const UserForm = (props) => {
  let history = useHistory();
  const { useCaseFactory } = useContext(AppContext);

  const [pageState, setPageState] = useState(
    new User({ id: undefined, name: "", isAdmin: false })
  );

  useEffect(() => {
    if (props.userId) {
      setPageState(
        new User({ id: props.userId, name: props.name, isAdmin: props.isAdmin })
      );
    }
  }, [props]);

  const onSubmit = async (user) => {
    let useCase;
    if (props.userId === undefined) {
      useCase = useCaseFactory.get(CreateUserUseCase);
    } else {
      useCase = useCaseFactory.get(UpdateUserUseCase);
    }
    const result = await useCase.execute(user);
    if (result) {
      history.push('/users');
    }
  };

  return (
    <Formik
      enableReinitialize={true}
      initialValues={pageState}
      validationSchema={Yup.object({
        name: Yup.string().required("Required"),
      })}
      onSubmit={onSubmit}
    >
      <Form>
        <div className="field">
          <label className="label" htmlFor="name">
            Name
          </label>
          <div className="control">
            <Field name="name" type="text" data-testid="name" />
          </div>
          <p className="help is-danger" data-testid="userNameErrors">
            <ErrorMessage name="name" />
          </p>
        </div>

        <div className="field">
          <label className="label" htmlFor="password">
            Password
          </label>
          <div className="control">
            <Field name="password" type="password" data-testid="password" />
          </div>
        </div>

        <div className="field">
          <label className="label" htmlFor="confirmPassword">
            Confirm Password
          </label>
          <div className="control">
            <Field
              name="confirmPassword"
              type="password"
              data-testid="confirmPassword"
            />
          </div>
        </div>

        <div className="field">
          <label className="label" htmlFor="isAdmin">
            Is Admin
          </label>
          <div className="control">
            <Field name="isAdmin" type="checkbox" data-testid="isAdmin" />
          </div>
        </div>

        <div className="field is-grouped">
          <p className="control">
            <button className="button" data-testid="submit" type="submit">
              {props.userId ? "UPDATE" : "CREATE"}
            </button>
          </p>
          <p className="control">
            <button
              className="button"
              data-testid="cancel"
              type="button"
              onClick={() => history.push("/users")}
            >
              CANCEL
            </button>
          </p>
          {props.userId && (
            <p className="control">
              <button
                className="button is-danger"
                data-testid="delete"
                type="button"
                onClick={() => history.push(`/users/${props.userId}/delete`)}
              >
                DELETE
              </button>
            </p>
          )}
        </div>
      </Form>
    </Formik>
  );
};
