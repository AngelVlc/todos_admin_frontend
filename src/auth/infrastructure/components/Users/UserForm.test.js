import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { UserForm } from "./UserForm";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { MemoryRouter, Route } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { CreateUserUseCase } from "../../../application/users";
import { User } from "../../../domain";

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const mockedCreateUserUseCase = {
  execute: jest.fn(),
};

const mockedUpdateUserUseCase = {
  execute: jest.fn(),
};

const useCaseFactory = {
  get: (useCase) => {
    if (useCase == CreateUserUseCase) {
      return mockedCreateUserUseCase;
    }

    return mockedUpdateUserUseCase;
  },
};

const renderWithContextAndRouterForExistingUser = (isAdmin) => {
  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <MemoryRouter initialEntries={[`/users/2/edit`]}>
          <Route path="/users/:userId/edit">
            <UserForm user={new User({ id: 2, name: "user", isAdmin })} />
          </Route>
        </MemoryRouter>
      </AppContext.Provider>
    ),
  };
};

const renderWithContextAndRouterForNewUser = () => {
  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <MemoryRouter initialEntries={[`/users/new`]}>
          <Route path="/users/new">
            <UserForm user={User.createEmpty()}/>
          </Route>
        </MemoryRouter>
      </AppContext.Provider>
    ),
  };
};

afterEach(cleanup);

it("should match the snapshot for an existing non admin user", async () => {
  let fragment;
  await act(async () => {
    const { asFragment } = renderWithContextAndRouterForExistingUser(false);
    fragment = asFragment;
  });
  expect(fragment()).toMatchSnapshot();
});

it("should match the snapshot for an existing admin user", async () => {
  let fragment;
  await act(async () => {
    const { asFragment } = renderWithContextAndRouterForExistingUser(true);
    fragment = asFragment;
  });
  expect(fragment()).toMatchSnapshot();
});

it("should match the snapshot for a new user", async () => {
  const { asFragment } = renderWithContextAndRouterForNewUser();
  expect(asFragment()).toMatchSnapshot();
});

it("should allow delete an existing user", async () => {
  let container;
  await act(async () => {
    container = renderWithContextAndRouterForExistingUser(false);
  });

  await waitFor(() => {
    fireEvent.click(container.getByTestId("delete"));
  });

  expect(mockHistoryPush.mock.calls.length).toBe(1);
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/users/2/delete");
  mockHistoryPush.mockClear();
});

it("should allow cancel", async () => {
  const { getByTestId } = renderWithContextAndRouterForNewUser();

  await waitFor(() => {
    fireEvent.click(getByTestId("cancel"));
  });

  expect(mockHistoryPush.mock.calls.length).toBe(1);
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/users");
  mockHistoryPush.mockClear();
});

it("should require user name", async () => {
  const { getByTestId } = renderWithContextAndRouterForNewUser();

  await waitFor(() => {
    fireEvent.click(getByTestId("submit"));
  });

  expect(getByTestId("userNameErrors")).toHaveTextContent("Required");
});

it("should update an existing user", async () => {
  let container;
  await act(async () => {
    container = renderWithContextAndRouterForExistingUser(false);
  });

  await changeInputValue(container.getByTestId, "name", "updated user");
  await changeInputValue(container.getByTestId, "password", "pass");
  await changeInputValue(container.getByTestId, "confirmPassword", "pass");
  await waitFor(() => {
    fireEvent.click(container.getByTestId("isAdmin"));
  });

  mockedUpdateUserUseCase.execute.mockResolvedValue({ id: 2 });

  await waitFor(() => {
    fireEvent.click(container.getByTestId("submit"));
  });

  expect(mockedUpdateUserUseCase.execute.mock.calls.length).toBe(1);

  const user = new User({ id: 2, name: "updated user", isAdmin: true });
  user.password = "pass";
  user.confirmPassword = "pass";
  expect(mockedUpdateUserUseCase.execute.mock.calls[0][0]).toStrictEqual(user);

  expect(mockHistoryPush.mock.calls.length).toBe(1);
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/users");
  mockHistoryPush.mockClear();
});

it("should create a new user", async () => {
  const { getByTestId } = renderWithContextAndRouterForNewUser();

  await changeInputValue(getByTestId, "name", "new user");
  await changeInputValue(getByTestId, "password", "pass");
  await changeInputValue(getByTestId, "confirmPassword", "pass");

  mockedCreateUserUseCase.execute.mockResolvedValue({ id: 55 });

  await waitFor(() => {
    fireEvent.click(getByTestId("submit"));
  });

  expect(mockedCreateUserUseCase.execute.mock.calls.length).toBe(1);

  const user = new User({ name: "new user", isAdmin: false });
  user.password = "pass";
  user.confirmPassword = "pass";
  expect(mockedCreateUserUseCase.execute.mock.calls[0][0]).toStrictEqual(user);

  expect(mockHistoryPush.mock.calls.length).toBe(1);
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/users");
  mockHistoryPush.mockClear();
});

const changeInputValue = async (getByTestId, name, value) => {
  await waitFor(() => {
    fireEvent.change(getByTestId(name), {
      target: {
        value: value,
      },
    });
  });
};
