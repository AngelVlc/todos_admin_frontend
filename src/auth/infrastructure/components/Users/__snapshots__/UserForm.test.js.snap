// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot for a new user 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="name"
      >
        Name
      </label>
      <div
        class="control"
      >
        <input
          data-testid="name"
          name="name"
          type="text"
          value=""
        />
      </div>
      <p
        class="help is-danger"
        data-testid="userNameErrors"
      />
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="password"
      >
        Password
      </label>
      <div
        class="control"
      >
        <input
          data-testid="password"
          name="password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="confirmPassword"
      >
        Confirm Password
      </label>
      <div
        class="control"
      >
        <input
          data-testid="confirmPassword"
          name="confirmPassword"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="isAdmin"
      >
        Is Admin
      </label>
      <div
        class="control"
      >
        <input
          data-testid="isAdmin"
          name="isAdmin"
          type="checkbox"
          value="false"
        />
      </div>
    </div>
    <div
      class="field is-grouped"
    >
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="submit"
          type="submit"
        >
          CREATE
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="cancel"
          type="button"
        >
          CANCEL
        </button>
      </p>
    </div>
  </form>
</DocumentFragment>
`;

exports[`should match the snapshot for an existing admin user 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="name"
      >
        Name
      </label>
      <div
        class="control"
      >
        <input
          data-testid="name"
          name="name"
          type="text"
          value="user"
        />
      </div>
      <p
        class="help is-danger"
        data-testid="userNameErrors"
      />
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="password"
      >
        Password
      </label>
      <div
        class="control"
      >
        <input
          data-testid="password"
          name="password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="confirmPassword"
      >
        Confirm Password
      </label>
      <div
        class="control"
      >
        <input
          data-testid="confirmPassword"
          name="confirmPassword"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="isAdmin"
      >
        Is Admin
      </label>
      <div
        class="control"
      >
        <input
          data-testid="isAdmin"
          name="isAdmin"
          type="checkbox"
          value="true"
        />
      </div>
    </div>
    <div
      class="field is-grouped"
    >
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="submit"
          type="submit"
        >
          UPDATE
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="cancel"
          type="button"
        >
          CANCEL
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button is-danger"
          data-testid="delete"
          type="button"
        >
          DELETE
        </button>
      </p>
    </div>
  </form>
</DocumentFragment>
`;

exports[`should match the snapshot for an existing non admin user 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="name"
      >
        Name
      </label>
      <div
        class="control"
      >
        <input
          data-testid="name"
          name="name"
          type="text"
          value="user"
        />
      </div>
      <p
        class="help is-danger"
        data-testid="userNameErrors"
      />
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="password"
      >
        Password
      </label>
      <div
        class="control"
      >
        <input
          data-testid="password"
          name="password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="confirmPassword"
      >
        Confirm Password
      </label>
      <div
        class="control"
      >
        <input
          data-testid="confirmPassword"
          name="confirmPassword"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="isAdmin"
      >
        Is Admin
      </label>
      <div
        class="control"
      >
        <input
          data-testid="isAdmin"
          name="isAdmin"
          type="checkbox"
          value="false"
        />
      </div>
    </div>
    <div
      class="field is-grouped"
    >
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="submit"
          type="submit"
        >
          UPDATE
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button"
          data-testid="cancel"
          type="button"
        >
          CANCEL
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button is-danger"
          data-testid="delete"
          type="button"
        >
          DELETE
        </button>
      </p>
    </div>
  </form>
</DocumentFragment>
`;
